udev: remove obsolete NVMe symlinks

bgurney-rh opened the pull request:
NVMe devices with namespace management capability can create multiple namespaces per controller, which can result in a condition where udev creates "/dev/disk/by-id" symbolic links that may "move around" between the namespaces of the NVMe storage device.

For an example of an NVMe storage device with five namespaces, this would be the output of "ls -l /dev/disk/by-id" symlinks:

nvme-MODEL_SERIAL -> ../../nvme1n1
nvme-MODEL_SERIAL_1 -> ../../nvme1n1
nvme-MODEL_SERIAL_2 -> ../../nvme1n2
nvme-MODEL_SERIAL_3 -> ../../nvme1n3
nvme-MODEL_SERIAL_4 -> ../../nvme1n4
nvme-MODEL_SERIAL_5 -> ../../nvme1n5

...and the "nvme-MODEL_SERIAL" symlink would rotate around between the five namespaces, as writes are performed to the namespaces:

nvme-MODEL_SERIAL -> ../../nvme1n4
...
nvme-MODEL_SERIAL -> ../../nvme1n2
...
nvme-MODEL_SERIAL -> ../../nvme1n5

(The "_1", "_2", etc. symlinks would remain pointing at the correct block device: the namespace that was identified as part of those udev rules.)

Therefore, stop creating this by-id namespace symlink.











<!-- devel-freezer = {"comment-id":"2125416251","freezing-tag":"v256-rc2"} -->

bluca commented:
The comment says `kept for backward compatibility` - so is this breaking backward compatibility? With what scenario?

johnmeneghini reviewed (approved):
This is an issue that has been reported by several of our customers here at Red Hat.  We discussed this issue at LSFMM and it was agreed that this and other bugs in udev needed to be fixed.

poettering commented:
Given this is kinda a compat break, could you please add a note about this change in NEWS already?

This is material for v257, hence make sure to add a new skeleton section for v257 on top of the file, with just a new entry indicating this compat break.

poettering commented:
also, can you add some comment to the commit msg (and the NEWS entry) indicating what the outcome of this is, i.e. what this means for users, and why it should be ok to change this?

bluca commented:
> The comment says `kept for backward compatibility` - so is this breaking backward compatibility? With what scenario?

And especially we need to understand precisely what scenario is affected and how

poettering commented:
another question: wouldn't it make sense to keep the symlinks but to automatically derive the priority for the symlink from the namespace id (that's a numeric value, right?)

```
…,OPTIONS+="link_priority=-$env{ID_NSID}",…
```

with something like that the namespace with the lowest ID will get the symlink?

to me it appears this would nicely retain compat but make behaviour systematic.

bgurney-rh commented:
> another question: wouldn't it make sense to keep the symlinks but to automatically derive the priority for the symlink from the namespace id (that's a numeric value, right?)
> 
> ```
> …,OPTIONS+="link_priority=-$env{ID_NSID}",…
> ```
> 
> with something like that the namespace with the lowest ID will get the symlink?
> 
> to me it appears this would nicely retain compat but make behaviour systematic.

Well, I tried that on the system with the drive that can reproduce the problem (based on systemd-254.10-1.fc39.x86_64), but I still saw behavior where either the symlink would still "move around" based on writes to the namespaces, or the link would not be created if I removed it.

(I wasn't sure which line should get the "link_priority" option, so I tried a few things:)

```
# obsolete symlink with non-escaped characters, kept for backward compatibility
ENV{ID_MODEL}=="?*", ENV{ID_SERIAL_SHORT}=="?*", ENV{ID_MODEL}!="*/*", ENV{ID_SERIAL_SHORT}!="*/*", \
  ENV{ID_SERIAL}="$env{ID_MODEL}_$env{ID_SERIAL_SHORT}", OPTIONS+="link_priority=-$env{ID_NSID}", SYMLINK+="disk/by-id/nvme-$env{ID_SERIAL}$env{.PART_SUFFIX}"
# obsolete symlink that might get overridden on adding a new nvme controller, kept for backward compatibility
ENV{ID_MODEL}=="?*", ENV{ID_SERIAL_SHORT}=="?*", OPTIONS="string_escape=replace", \
  ENV{ID_SERIAL}="$env{ID_MODEL}_$env{ID_SERIAL_SHORT}", SYMLINK+="disk/by-id/nvme-$env{ID_SERIAL}$env{.PART_SUFFIX}"


(link target still migrates around to last namespace written)




# obsolete symlink with non-escaped characters, kept for backward compatibility
ENV{ID_MODEL}=="?*", ENV{ID_SERIAL_SHORT}=="?*", ENV{ID_MODEL}!="*/*", ENV{ID_SERIAL_SHORT}!="*/*", \
  ENV{ID_SERIAL}="$env{ID_MODEL}_$env{ID_SERIAL_SHORT}", SYMLINK+="disk/by-id/nvme-$env{ID_SERIAL}$env{.PART_SUFFIX}"
# obsolete symlink that might get overridden on adding a new nvme controller, kept for backward compatibility
ENV{ID_MODEL}=="?*", ENV{ID_SERIAL_SHORT}=="?*", OPTIONS="string_escape=replace", OPTIONS+="link_priority=-$env{ID_NSID}", \
  ENV{ID_SERIAL}="$env{ID_MODEL}_$env{ID_SERIAL_SHORT}", SYMLINK+="disk/by-id/nvme-$env{ID_SERIAL}$env{.PART_SUFFIX}"

(link target still migrates around to last namespace written)



# obsolete symlink with non-escaped characters, kept for backward compatibility
ENV{ID_MODEL}=="?*", ENV{ID_SERIAL_SHORT}=="?*", ENV{ID_MODEL}!="*/*", ENV{ID_SERIAL_SHORT}!="*/*", \
  ENV{ID_SERIAL}="$env{ID_MODEL}_$env{ID_SERIAL_SHORT}", OPTIONS+="link_priority=-$env{ID_NSID}", SYMLINK+="disk/by-id/nvme-$env{ID_SERIAL}$env{.PART_SUFFIX}"
# obsolete symlink that might get overridden on adding a new nvme controller, kept for backward compatibility
# ENV{ID_MODEL}=="?*", ENV{ID_SERIAL_SHORT}=="?*", OPTIONS="string_escape=replace", \
#   ENV{ID_SERIAL}="$env{ID_MODEL}_$env{ID_SERIAL_SHORT}", SYMLINK+="disk/by-id/nvme-$env{ID_SERIAL}$env{.PART_SUFFIX}"

(link target doesn't change around; link won't be created.)



# obsolete symlink with non-escaped characters, kept for backward compatibility
# ENV{ID_MODEL}=="?*", ENV{ID_SERIAL_SHORT}=="?*", ENV{ID_MODEL}!="*/*", ENV{ID_SERIAL_SHORT}!="*/*", \
#   ENV{ID_SERIAL}="$env{ID_MODEL}_$env{ID_SERIAL_SHORT}", SYMLINK+="disk/by-id/nvme-$env{ID_SERIAL}$env{.PART_SUFFIX}"
# obsolete symlink that might get overridden on adding a new nvme controller, kept for backward compatibility
ENV{ID_MODEL}=="?*", ENV{ID_SERIAL_SHORT}=="?*", OPTIONS="string_escape=replace", OPTIONS+="link_priority=-$env{ID_NSID}", \
  ENV{ID_SERIAL}="$env{ID_MODEL}_$env{ID_SERIAL_SHORT}", SYMLINK+="disk/by-id/nvme-$env{ID_SERIAL}$env{.PART_SUFFIX}"

(link target doesn't change around; link won't be created.)
```

It's late in the day, so I may have lost track of things, but there was one configuration that I tried, where the correct namespace links (i.e.: "..._1", "..._2", etc.) started vanishing.  I forget how I did that, but I reverted it on the test system, and tried the changes listed above.

bgurney-rh commented:
I drafted a NEWS addition; I'm worried that the wording might need to be clarified, so I'll add what I have composed so far in this comment:
```
CHANGES WITH 257:

        * udev now only creates symlinks to the namespaces of PCI Express NVMe
          drives using the "disk/by-id/nvme-MODEL_SERIAL_NS" convention.  The
          "dev/disk/by-id/nvme-MODEL_SERIAL" symlink would work properly for 
          drives with only one namespace (in other words, without the
          capability of namespace management), but for drives with the
          capability of providing multiple namespaces, the target of this
          symlink would change between the namespaces of the drive, based on
          write activity.  For this reason, the symlinks with the namespace
          identifier (i.e.: "nvme-MODEL_SERIAL_NS") should be considered more
          reliable (however, a more reliable udev symlink would be the
          "disk/by-uuid" link, based on written metadata headers of the
          filesystem or storage subsystem, or the hardware identifier of the
          drive; i.e.: eui64, nguid, etc.).  For more details: see
          https://github.com/systemd/systemd/issues/31448
```

When I gave the presentation at LSF/MM, and demonstrated the "moving around" behavior of the target of the symlink with only the drive's model and serial number, the mood in the room was very much that of, "We should stop using this symlink."  Most earlier PCI Express NVMe drives did not support namespace management, and therefore could only present one namespace.  Drives that can present multiple namespaces are starting to become more available on the market, so more users could encounter this problem if the unreliable symlink is still created.

(Also, I'm guessing that there's a CI test failure due to removing a symlink that was previously expected to be there, but this commit would be removing; I don't know how to edit that test yet, but I can add that in as well.)

bluca commented:
> The comment says `kept for backward compatibility` - so is this breaking backward compatibility? With what scenario?

Could you please clarify exactly what compatibility is being broken, and with what types of devices? Where are they used? Etc. Breaking compatibility is not something we do lightly.

yuwata commented:
> > What happens if the old-style symlink is explicitly used, e.g. fstab or kernel command line option? I guess the change easily makes the system unbootable. So, even if they are meaningless, I think we should keep them. That's the meaning of keeping backward compatibility. Unless the old-style symlinks cause some significant issues we cannot avoid, we should keep them.
> 
> I agree, and looking at how red the CI is doesn't exactly tell me this is safe to do. Maybe they could be split in a separate rule file by themselves, so that distributions/users can easily override/mask them?

Yeah, sounds good idea. How about to introduce 61-persistent-storage-compat.rules or so, and move the compat symlinks there?

bluca commented:
> > > What happens if the old-style symlink is explicitly used, e.g. fstab or kernel command line option? I guess the change easily makes the system unbootable. So, even if they are meaningless, I think we should keep them. That's the meaning of keeping backward compatibility. Unless the old-style symlinks cause some significant issues we cannot avoid, we should keep them.
> > 
> > 
> > I agree, and looking at how red the CI is doesn't exactly tell me this is safe to do. Maybe they could be split in a separate rule file by themselves, so that distributions/users can easily override/mask them?
> 
> Yeah, sounds good idea. How about to introduce 61-persistent-storage-compat.rules or so, and move the compat symlinks there?

That sounds good to me

poettering commented:
So yeah, I too thin that it's one thing if the symlink as it is right now doesn't make sense on newer devices. It's another thing that people might actually use the symlink already and we shouldn't willy-nilly break it. Breaking things might be OK from time to time but that requires a high bar of breakage really. In particular as I understand you most devices in the wild to this day only have a single namespace, no, and using the symlnk there is not a problem, or is it?

I think the link priority stuff would be much nicer, since it should make the symlink – as weakly as it is defined – something systematic, i.e. should result in always the "same" namespace device being referenced by the symlink, if multiple namespaces are defined, no?

yuwata commented:
Note, `link_priority=` is per-device, instead of per-symlink.

Moreover, it does not support substitution, so e.g. `link_priority=$attr{foo}` does not work, at least currently.

keszybz commented:
> Note, `link_priority=` is per-device, instead of per-symlink.

That's OK, no? We have different devices in different namespaces
and we pick one of them to symlink to.

> Moreover, it does not support substitution, so e.g. `link_priority=$attr{foo}` does not work, at least currently.

This would explain why this didn't work. It shouldn't be too hard to fix.

johnmeneghini commented:
On 5/28/24 13:36, Zbigniew Jędrzejewski-Szmek wrote:
>  > Note, `link_priority=` is per-device, instead of per-symlink.
> 
> That's OK, no? We have different devices in different namespaces
> and we pick one of them to symlink to.
> 
>  > Moreover, it does not support substitution, so e.g. `link_priority=$attr{foo}` does not work, at least currently.
> 
> This would explain why this didn't work. It shouldn't be too hard to fix.


So what's your proposed fix? Keep the simlinks in /dev/disk/by-path?  Or are we talking about the /dev/disk/by-id?

/John

> —
> Reply to this email directly, view it on GitHub <https://github.com/systemd/systemd/pull/32981#issuecomment-2135782891>, or 
> unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ADCUXXL5GMUQMR3JQWROR4TZES6AXAVCNFSM6AAAAABIEDISVCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMZVG44DEOBZGE>.
> You are receiving this because you commented.Message ID: ***@***.***>
>

yuwata commented:
Sorry if you are confused by https://github.com/systemd/systemd/pull/32981#issuecomment-2135660943.

I just noted the current status of link_priority option, literally.

> > Moreover, it does not support substitution, so e.g. `link_priority=$attr{foo}` does not work, at least currently.
> This would explain why this didn't work. It shouldn't be too hard to fix.

Right. It is easy to extend the feature.

> So what's your proposed fix? Keep the simlinks in /dev/disk/by-path?  Or are we talking about the /dev/disk/by-id?

Please keep all compat symlinks.

bgurney-rh commented:
I'd like to invite @keithbusch and @hreinecke to comment and review on this, as they are the kernel nvme maintainers who expressed concerns about the "compat symlinks" being invalid, when I gave the presentation at LSF/MM two weeks ago.

I haven't tested it yet, but there might be a way to create a reproducer virtual machine via qemu (see "Additional Namespaces" in https://qemu-project.gitlab.io/qemu/system/devices/nvme.html ).

keithbusch commented:
> I'd like to invite @keithbusch and @hreinecke to comment and review on this, as they are the kernel nvme maintainers who expressed concerns about the "compat symlinks" being invalid, when I gave the presentation at LSF/MM two weeks ago.
> 
> I haven't tested it yet, but there might be a way to create a reproducer virtual machine via qemu (see "Additional Namespaces" in https://qemu-project.gitlab.io/qemu/system/devices/nvme.html ).

I can't imagine the link without a namespace instance appended was anything but a mistake. The spec always allowed multi-namespace devices; the very first nvme device I ever used had 128 namespaces (never shipped). They just needed to be static. So that link was always broken, it was never reliable. I recall I initially submitted the udev rules for nvme's by-path links, but I don't recall this one and I think it should be deleted because it is completely unreliable and leads to inevitable data corruption.

hreinecke commented:
Couldn't agree more. Since the very beginning nvme (block) devices _always_ were using the namespaces (ie nvmeXnY); the controller device (nvmeX) was always a character control device and should never have been listed here.
That was a mistake and should never have happened. I doubt whether this qualifies as compability issue; it's just plain wrong.

yuwata commented:
Again, even if the concept of the compat symlinks is messy, we cannot remove them easily, unless they cause some serious issues.

Or, do you have any real issues caused by the compat symlinks?

poettering commented:
I googled a bit, and found various docs that suggest the names without a namespace suffix, for example this here:

https://www.ibm.com/docs/en/linux-on-systems?topic=drivers-nvme
https://docs.netapp.com/us-en/e-series/config-linux/nvme-fc-access-physical-targets-concept.html
https://thinksystem.lenovofiles.com/storage/help/index.jsp?topic=%2Fde-series-11.70.4_config-linux%2Fnvme-ib-access-physical-targets-concept.html

Hence, mistake or not, it appears the names by-id names lacking a namespace suffix are clearly well documented, we cannot just ignore that. I think we have two options:

1. always make them point to namespace 1 of the device. If there is no namespace 1 the link would not exist.
2. do the link priority thing, so that the lowest namespace that exists gets the link.

I don't really care much which approach is used here.

hreinecke commented:
I would vote for case 1).
NVMe devices without the namespace number are controller devices, which are character devices, not block devices.

keithbusch commented:
> I googled a bit, and found various docs that suggest the names without a namespace suffix, for example this here:
> 
> https://www.ibm.com/docs/en/linux-on-systems?topic=drivers-nvme https://docs.netapp.com/us-en/e-series/config-linux/nvme-fc-access-physical-targets-concept.html https://thinksystem.lenovofiles.com/storage/help/index.jsp?topic=%2Fde-series-11.70.4_config-linux%2Fnvme-ib-access-physical-targets-concept.html

The IBM doc is just blatantly incorrect. The "n#" has nothing to do with the "namespace id" assigned by the disk. It's just a unique number assigned by the operating system.

The other two docs only refer to by-id links using EUI naming and those are fine as-is without a namespace suffix because they are already unique per namespace. They have nothing to do with this issue.

> always make them point to namespace 1 of the device. If there is no namespace 1 the link would not exist.
> do the link priority thing, so that the lowest namespace that exists gets the link.

How do you do this with dynamic namespaces? If you detach the namespace the symlink currently points to, does it get reassigned? And if I attach a new namespace with a lower identifier, does that steal the symlink from whoever currently is using it? This a data corruption trap.

bgurney-rh commented:
What happens if the only namespace presented by the NVMe device isn't "namespace 1", but some other number?

poettering commented:
> What happens if the only namespace presented by the NVMe device isn't "namespace 1", but some other number?

As I wrote earlier: no symlink without namespace suffix would be generated in that case. Which is totally fine.

poettering commented:
> Then there are no symlinks… That may be a problem, but the kernel seems to pick 1 in general. 

As I understand it's not the kernel that picks that but the driver and the admin chose that.

> If we accept the requirement that the symlink must not jump between devices, then we need to make it stable, and making it always point to 1 is the only reasonable way of making it stable.

Also, let's not forget *why* we even care: for compat with existing systems. And those typically only have one namespace, and it's numbered one. Hence, if twe create the compat symlinks only for the namespace #1 we should provide all compat we need. And people who have more than one namespace shouldn't use these compat symlinks anyway, but the ones that carry a namespace suffix.

keithbusch commented:
> > Then there are no symlinks… That may be a problem, but the kernel seems to pick 1 in general.
> 
> As I understand it's not the kernel that picks that but the driver and the admin chose that.

I want to clarify how the driver naming works because I'm not sure it's clear.

Frequently asked question: Why does the kernel assign a number, '`Y`', to the "`nvmeXnY`" handle instead of using the already unique Namespace ID provided by the device?

Answer: User space may be holding a reference against a kernel handle (ex: a stale mount point), and dynamic namespace management may detach and attach a different namespace with the same number. Attempting to surface the newly attached namespace will create naming collisions within the kernel.

So if you want to append the namespace ID and ensure it's always assigned to NSID 1? That may or may not align to `/dev/nvmeXn1`, because the "`n1`" may refer to something else. Linking to it may or may not refer to the same block device on any given boot.

Does this matter? I suspect "NO"; these capabilities I'm describing are in the enterprise space.

poettering commented:
Still, we shouldn't just drop the old symlinks, people use them.

I think the simplest solution is to point them to namespace 1, if there is one, and if there is none to not create them. I think this should provide us with enough compat, as old setups will only have namespace 1. And everything is very clear then. Can you rework the patch like that?

i guess all that's necessary for that is to include an additional `ENV{ID_NSID}=="1"` match in those rules.

Then we get the best of all worlds: no ambiguity and enough compat.

lnykryn commented:
Honestly, we should not fix it. If the naming is inherently wrong, we should do the same thing we are doing for NIC naming. Lock the old naming behind some kernel cmdline option (maybe we can do a general udev.naming=v256 both for storage and networking) and remove this for newer versions.

bgurney-rh commented:
Okay, I tested out the additional `ENV{ID_NSID}=="1"` match, and it seems to work properly for multiple-namespace drives.  I'm going to create a new pull request with that change, to be posted soon.

poettering commented:
I guess this can be closed now that #34482 has been merged,
